### DISCUSSION TITLE
[Feature Request] [Monarch] Support for End of File/Input or Way to determine if line being tokenized is last line or not

### DISCUSSION THREAD

**Discussion by akshar2401** [microsoft/monaco-editor#3991] [Q&A]
### Context

- [X] This issue is not a bug report. *(please use a different template for reporting a bug)*
- [X] This issue is not a duplicate of an existing issue. *(please use the [search](https://github.com/microsoft/monaco-editor/issues) to find existing issues)*

### Description

Hi,
I have a use case for the custom language i am using monaco-editor for. I want to be able to detect whether a certain rule/regex in Monarch configuration ends in the last line (end of file/input). For example, I configured monarch to tokenize mulitline strings such as "Line 1\nLine2". I want to tokenize whole string as "Error" if it is not terminated on the last line or at the end of file. Currently, "$" is used for end of source line as monaco performs line by line tokenization. Adding a link to the playground. Using "$" only works when the string is just single line. May be a special operator to indicate whether line being tokenized is last or EOF

### Monaco Editor Playground Link

https://microsoft.github.io/monaco-editor/playground.html?source=v0.38.0#XQAAAAIGAwAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw2yEzSAEiArd-Hl99HCj96EBbem_SwwMNaAADDdo9qw1c65aJytU15iTreXI4n3HI2iltrf1qjgNxT39SiQ7_hBVxCo7QeRAkoi6304bmAk_UHNAwGcwV-RQnhWYhyjsP75G7Caqb5lPnwPThMY7Mk9zbrs1ybpsoVUhsVcBqSIxJaBnh15bnpx9MbJNGvecYgsKEdROzl-aLDsJBlYM8SSMHk4bTLNIADi-O8SLt0-kXCWHrF9F2BkgGyj3IiPxAW9IIfuTq_cdRw4BL7669KLdz5A8uc-pm56bICh1vKFyZjFk6Zgs4sVnnvL4AK1QdshOSIeWUyovikQtKutd4wOPlpsoogkDi2gKETWcBMuI0R0C2HQakKq4Z5hJtQkEOXZ3FDQYsw88t64V8bUAZWhJSki6nv513oHzMvJrf6xHNEWjfdMKwabgqwqLv9ea159j_a8xN3ZL6uG7GOzpXcbEGXwdfyJYGmOmz8EIWwWD45btfgyeYkcYy67Htv8T_SyjZkCI8LBbbOAxpm5HibsmOoZs2t2xiP__DDbA_

### Monaco Editor Playground Code

```typescript
const lang = "testlan"
monaco.languages.register({id: lang})
monaco.languages.setMonarchTokensProvider(lang, {
	tokenizer: {
		root: [
			{include: "String"}
		],
		String: [
			[/"/, {token: 'string', next: 'String.Body'}],
		],
		'String.Body': [
			// If string is unterminated
			[/([^"]|(""))+$/gu, {token: "Error"}],
			[/([^"]|(""))+/gu, {token: "string", switchTo: 'String.Body'}],
			[/"/, "string", "@pop"],
		]
	}
})
monaco.editor.defineTheme('Theme', {
	base: 'vs',
	inherit: true,
	colors: {},
	rules: [
		{token: 'string', foreground:'fc1703'}
	]
	
})
const editor = monaco.editor.create(document.getElementById("container"), {
	language: lang,
	theme: "Theme",
	value: `"Line 
	1 Line 2
	`
})
```

**Comment by hediet** [comment]
I don't think this is supported. For diagnostics, you should use the diagnostic API and not tokens.

**Reply by akshar2401** [reply]
I am definitely using diagnostics API but I have a tokenization use case in this. I color the string orange but I want the change the color to black if string is not terminated on the last line of the content in the editor. When this is the case, error markers are shown but the string is still orange. I want to change the color to black.

**Reply by akshar2401** [reply]
I saw an option “includeLF” which adds the “\n” to each line being tokenized. It is also added to the last line. Not sure if it makes sense but May be “\n” shouldn’t be added to the last line as there is no line after the last line. I am not sure if it would work or not but I can use the presence or absence of “\n” to solve my use case
